### PR TITLE
Load test

### PR DESCRIPTION
**Pull request by tirsen**

### CODE REVIEW

File: pkg/clone/checksum.go [hunk 1]
```diff
@@ -273,3 +319,223 @@ func (cmd *Checksum) run(ctx context.Context) ([]Diff, error) {
...
+			return nil
+		}
+		if sourceRow != nil && targetRow == nil {
+			newDiff = &Diff{
+				Type:   Insert,
+				Row:    sourceRow,
+				Target: nil,
+			}
+			return nil
+		}
```
Comment (tirsen) [inline pkg/clone/checksum.go:423]: if both rows are == nil then it should be a no diff

File: pkg/clone/checksum.go [hunk 2]
```diff
@@ -273,3 +319,223 @@ func (cmd *Checksum) run(ctx context.Context) ([]Diff, error) {
...
+		if rowsEqual {
+			newDiff = nil
+			return nil
+		} else {
+			newDiff = &Diff{
+				Type:   Update,
+				Row:    sourceRow,
+				Target: targetRow,
+			}
+			return nil
```
Comment (tirsen) [inline pkg/clone/checksum.go:437]: delete this row

File: pkg/clone/checksum.go [hunk 3]
```diff
@@ -273,3 +319,223 @@ func (cmd *Checksum) run(ctx context.Context) ([]Diff, error) {
...
+		}
+		if sourceRow != nil && targetRow == nil {
+			newDiff = &Diff{
+				Type:   Insert,
+				Row:    sourceRow,
+				Target: nil,
+			}
+			return nil
```
Comment (tirsen) [inline pkg/clone/checksum.go:422]: delete this row

File: pkg/clone/checksum.go [hunk 4]
```diff
@@ -273,3 +319,223 @@ func (cmd *Checksum) run(ctx context.Context) ([]Diff, error) {
...
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		if sourceRow == nil && targetRow != nil {
+			newDiff = &Diff{
+				Type:   Delete,
+				Row:    targetRow,
+				Target: nil,
+			}
+			return nil
```
Comment (tirsen) [inline pkg/clone/checksum.go:414]: delete this row

File: pkg/clone/differ.go [hunk 5]
```diff
@@ -359,6 +360,7 @@ func (r *Reader) doDiffChunk(ctx context.Context, chunk Chunk) ([]Diff, error) {
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
+	targetStream.sort()
```
Comment (tirsen) [inline pkg/clone/differ.go:363]: leave comment here why I sort

File: pkg/clone/replicate.go [hunk 6]
```diff
@@ -176,6 +207,44 @@ func (cmd *Replicate) ReconnectBackoff() backoff.BackOff {
 	return b
 }
 
+func (cmd *Replicate) readLag(ctx context.Context) (time.Duration, error) {
```
Comment (tirsen) [inline pkg/clone/replicate.go:210]: no longer used

File: pkg/clone/snapshotter.go [hunk 7]
```diff
@@ -243,15 +247,36 @@ func (c *ChunkSnapshot) findRow(row []interface{}) (*Row, int, error) {
...
+
+func (c *ChunkSnapshot) sort() {
+	sort.Slice(c.Rows, func(i, j int) bool {
+		return c.Rows[i].PkLess(c.Rows[j].Data)
+	})
+}
+
 func (c *ChunkSnapshot) updateRow(i int, row []interface{}) {
+	//logrus.WithField("task", "snapshot").Infof(
+	//	"reconciling binlog event inside watermarks by UPDATE chunk=%s index=%d old=%v new=%v", c.Chunk.String(), i, c.Rows[i], row)
```
Comment (tirsen) [inline pkg/clone/snapshotter.go:267]: remove

File: pkg/clone/replicate.go [hunk 8]
```diff
@@ -157,7 +159,26 @@ func (cmd *Replicate) run(ctx context.Context) error {
...
-			time.Sleep(cmd.DoSnapshotDelay)
+			logger := logrus.WithField("task", "snapshot")
+			for {
+				delay := cmd.DoSnapshotMaxReplicationLagCheckInterval
+				lag := replicator.GetReplicationLag()
+				if err != nil {
+					logger.WithError(err).Warnf("failed to check replication lag, checking again in %v", delay)
+					continue
+				}
+				if lag < cmd.DoSnapshotMaxReplicationLag {
```
Comment (yyuan-squareup) [inline pkg/clone/replicate.go:170]: - [ ] if the lag is always >= cmd.DoSnapshotMaxReplicationLag, would it be an infinite loop?

**Review by yyuan-squareup — APPROVED** [review]
LGTM, left a TODO

File: pkg/clone/replicate.go (hunk 8 above)
Comment (tirsen) [reply pkg/clone/replicate.go:170]: yeah it wouldn't start a snapshot until the lag is low enough. I think that's good?

**Review by evjan — APPROVED** [review]

**Pull request merged**